### ISSUE TITLE
(ec2): default Vpc structure results in broken networking during create/delete

### ISSUE THREAD

**Issue by kylelaker** [aws/aws-cdk#21348]
### Describe the bug

When creating a VPC with the default configuration, in some situations, the NAT Gateway(s) may be torn down before the private subnets are. This can cause issues for resources that rely on network egress in order to successfully (such as a Lambda-backed CloudFormation Custom Resource). Additionally, the Private subnets (and resources that depend on them) may be created before the NAT Gateways. This can result in broken initialization logic.

### Expected Behavior

It should not be possible for the NAT Gateway to be deleted before the private subnets. The private subnets should depend on the Gateway.

### Current Behavior

The NAT Gateway resource, the public subnets, and the internet gateway are deleted.

### Reproduction Steps

```ts
import * as cdk from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as lambda from "aws-cdk-lib/aws-lambda";

const app = new cdk.App();
const stack = new cdk.Stack(app, "TestStack");
const vpc = new ec2.Vpc(stack, "TestVpc");

const fn = new lambda.Function(stack, "TestFn", {
  vpc,
  vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_NAT },
  // Other stuff to require internet access
});
```

Building out the necessary infrastructure from here to actually create a custom resource to demonstrate is somewhat non-trivial and would e able to be copied/pasted.

### Possible Solution

Whether directly or indirectly, the private subnet should depend on the NAT Gateway. This could also be done by depending on the route. This should be handled already in `configureSubnet`. https://github.com/aws/aws-cdk/blob/2d269165b56c3fb19a75138be9109ca7a44137ac/packages/%40aws-cdk/aws-ec2/lib/nat.ts#L248-L255

But the Route Table Association has the references to the Subnet and the Route Table, and the Routes reference the Route Table and the target (NAT Gateway). But there's no reference from the Subnet to any of those resources, so there is no implicit dependency, which breaks network egress for resources in private subnets as the stack is being deleted.

Maybe this issue is limited to Lambda Functions instead of being a larger problem with EC2? Do more resources need to add a dependency on `Vpc.internetConnectivityEstablished`?

### CDK CLI Version

2.33.0

### Node.js Version

16

### OS

Linux

### Language

Typescript

**Comment by corymhall** [comment]
@kylelaker it sounds like this may be an issue specific to CloudFormation custom resources since
those are the only types of resources that you would care about successfully executing while
*destroying* the stack.

I think the solution here would be to specify the dependency on `internetConnectivityEstablished`
within the lambda.

This issue has been classified as *p2*. That means a workaround is available or it is deemed a nice-to-have feature. Given the amount of work there is to do and the relative priority of this issue, the CDK team is unlikely to address it. That does not mean the issue will never be fixed! If someone from the community submits a PR to fix this issue, and the PR is small and straightforward enough, and meets the quality bars to be reviewed and merged with little effort we will accept that PR. PRs that do not build or need complex or multiple rounds of reviews are unlikely to be merged and will be closed to keep our backlog manageable. 

In the mean time, remember that you can always use the escape hatch (https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html) mechanism to have fine control over the CloudFormation output you want. We will keep the issue open for discoverability, to collect upvotes, and to facilitate discussion around this topic.

We use +1s *on this issue* to help prioritize our work, and are happy to re-evaluate the prioritization of this issue based on community feedback. You can reach out to the cdk.dev community on Slack (https://cdk.dev/) to solicit support for reprioritization.

**Comment by kylelaker** [comment]
So I think that's part of it. But there's something even weirder happening when specifying a dependency on `internetConnectivityEstablished` for a private subnet. For example:

```ts
#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as lambda from "aws-cdk-lib/aws-lambda";

const app = new cdk.App();
const stack = new cdk.Stack(app, "Ec2PrivateStack");
const vpc = new ec2.Vpc(stack, "SampleVpc");
const vpcSubnets = { subnetType: ec2.SubnetType.PRIVATE_WITH_NAT };

const fn = new lambda.SingletonFunction(stack, "Function", {
  code: lambda.Code.fromInline(`module.exports = { handler: function() { console.log("Hello, world"); } }`),
  runtime: lambda.Runtime.NODEJS_16_X,
  handler: "index.handler",
  uuid: "fa5088ab-850a-4106-a657-482d78dccf68",
  lambdaPurpose: "Hello",
  vpc,
  vpcSubnets,
});

fn.addDependency(vpc.selectSubnets(vpcSubnets).internetConnectivityEstablished);
```

So the dependency on the private routes gets set, but it seems like other dependencies aren't really happening in a way that results in success. It works for creation but then the stack fails to delete. You can see that the public subnet resources start to get deleted immediately. So the private subnet routes and the NAT Gateway may still exist but they have no routes to the IGW. And from there because the Lambda function takes some time to delete (well, it actually deletes pretty quickly but CloudFormation takes ~20min to notice but that feels like an unrelated CloudFormation bug) other resources keep deleting in weird orders and the whole process ends up failing with a lingering Internet Gateway and VPC with "dependent resources".

This means the given workaround does not work. And it makes #21357 pretty impossible to implement correctly.

I **think** the fix here would be to make the `CfnNatGateway` defined at https://github.com/aws/aws-cdk/blob/618c5bcb33ba4a1b82cedf794b699cbfb8d21b9c/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L1937-L1946 depend not only on the subnet (via the implicit dependency caused by the reference to its ID) but also to add an explicit dependency on `this.internetConnectivityEstablished`; otherwise every resource that needs internet connectivity in a `PRIVATE_WITH_NAT` subnet also has to add a dependency on the public subnets _that have NAT Gateways_ which is a hard filter to add.

The stack events from building and deleting this stack can be seen at https://gist.github.com/kylelaker/02113c74bef0fdf4e1d37885e65d688f. I think I ran that command a little to early so the final VPC delete failure isn't there but I think the point is clear.

I will open a PR to add the dependency to the NAT Gateway but let me know if that's not the right approach.

**Comment by rix0rrr** [comment]
I would need to see the dependency structure here to see where we can add a dependency line most conveniently without introducing a cycle.

I'm guessing right now `internetConnectivityEstablished` results in a dependency on the route tables... but I'm not exactly sure how the CFN resources are laid out.

**Comment by rix0rrr** [comment]
I believe it's this. Will continue tomorrow:

![image](https://user-images.githubusercontent.com/524162/183467674-20c943b6-22d6-4e44-b463-0f3dcd6c3eb8.png)

**Comment by kylelaker** [comment]
> I guess the safe solution is to have `Function` add the dependency to itself by default. I'm a little concerned about pessimizing deployment speed for Functions that _don't_ need to execute during the deployment (as basically only Custom Resources need to do this).

I tried to think through this to see if I could identify a better solution than doing this for _all_ `Function`s; but while this can be done for `Provider` and `AwsCustomResource` (well, that one only matters if #21357 is merged), there's still `CustomResourceProvider`/`CustomResource` in `core` (which are primarily used in cases where adding `VpcConfig` would be tricky anyway) that are less important.

But then I realized that theoretically a Lambda function that is used as a Custom Resource (and therefore would get the dependency via `Provider`) could end up invoking other Lambda functions (either directly or indirectly) that are _also_ then going to (potentially) require internet connectivity.

And it feels feasible to document "if you use a `Function` directly or indirectly as part of a Custom Resource, make sure to add a dependency on internet connectivity". But that seems like a lot of unnecessary copying and pasting (and potential uncertainty).

So I agree that just doing the safer thing by default feels most correct and is closer to what I'd expect as an end user.
